### PR TITLE
feat: add E2E tests for chromium and firefox

### PR DESCRIPTION
**Pull request by galargh**
###### Description

This PR adds a setup for E2E testing to `ipfs-companion`.

The environment setup is automated using docker compose. The environment consists of:
- firefox container
- chrome container
- kubo container
- e2e container where the tests are executed from

The tests check the following for firefox and chromium:
- can the extension be installed?
- can api/gateway URL be modified?
- is the number of connected peers correctly reported?
- is the number of connected peers greater than 0?

The tests follow the proposed testing procedure from https://github.com/ipfs/kubo/blob/master/docs/RELEASE_ISSUE_TEMPLATE.md.

The newly added tests can be executed through GitHub Actions.

###### Testing
- [x] build companion from source: https://github.com/galargh/ipfs-companion/actions/runs/3739328246
- [x] download companion from releases: https://github.com/galargh/ipfs-companion/actions/runs/3739280013

###### TODO

- [x] create GHA workflow for running E2E tests; proposed inputs
  - [x] firefox/chromium versions
  - [x] ipfs-companion version; proposed behaviours
      - [x] build from source
      - [x] download from releases
   - [x] kubo version
- [x] add description

### CODE REVIEW

File: ci/access-control-allow-all.sh
Comment (galargh) [reply ci/access-control-allow-all.sh:5]: Anything in particular we're concerned with? The reason why this is needed is because in the "everything in docker" setup, the browsers run in different containers than kubo. There might be another way to do it. This was just the one that required the least thought.

File: ci/download-release-artifacts.sh
Comment (galargh) [reply ci/download-release-artifacts.sh:13]: The problem with using an action for this is that it cannot be executed locally. I found it useful to be able to call `ci/download-release-artifacts.sh` and work with released artifacts.

File: docker-compose.e2e.yml
Comment (galargh) [reply docker-compose.e2e.yml:4]: There's a disclaimer in the `playwright/firefox` image:
> These images only work with Moon software (https://aerokube.com/moon/). Official Playwright team images are located here: https://hub.docker.com/_/microsoft-playwright

It looks like the browsers are embedded in the main playwright image though - https://playwright.dev/docs/docker. Less customisable but I think it might be OK for this purpose.

There's a bigger blocker however. It looks like loading Firefox extensions is not supported by playwright. See https://github.com/microsoft/playwright/issues/7297

File: package.json
Comment (galargh) [reply package.json:62]: `-it` doesn't work in no tty environments - e.g. on a GHA runner. Also, it seems to me that running `npm run ci:build` inside a docker container shouldn't require interactivity anyway.

File: package.json
Comment (galargh) [reply package.json:64]: I'm only adding a trailing comma here.

File: package.json
Comment (galargh) [reply package.json:65]: I tried `build --pull` and, unlike `pull`, it will not pull `firefox`, `chrome` and `kubo` images. I think it only pulls the images that are used in the building process of other images, `e2e` one in this case.

File: test/e2e/ipfs-companion.test.js
Comment (galargh) [reply test/e2e/ipfs-companion.test.js:6]: No particular reason to begin with but now I found that we cannot do the same in `playwright` due to firefox extensions not being supported there (https://github.com/microsoft/playwright/issues/7297).

File: test/e2e/ipfs-companion.test.js
Comment (galargh) [reply test/e2e/ipfs-companion.test.js:10]: I'm gonna apply the suggestion on `async`. Our linting rules require the reject part of a promise to be explicitly specified though.

```suggestion
function delay (ms) {
  return new Promise(res => setTimeout(res, ms), _ => {})
}
```

File: test/e2e/ipfs-companion.test.js
Comment (galargh) [reply test/e2e/ipfs-companion.test.js:14]: I did try to import `add-on/manifest.common.json` first but this required an `assert` as in:
```
import { version } from '../../../add-on/manifest.common.json` assert { type: 'json' } 
```

And the linter's parser wasn't happy about it. It was failing to parse a file altogether.

File: test/e2e/ipfs-companion.test.js
Comment (galargh) [reply test/e2e/ipfs-companion.test.js:30]: I don't quite understand this one. Also, FYI, the linter would complain about `!=` not being `!==`.

File: test/e2e/ipfs-companion.test.js
Comment (galargh) [reply test/e2e/ipfs-companion.test.js:36]: That's selenium and what I saw when I was running this is that it took a while for the extension's `index` page to load and until that happened we weren't able to find its tab/extension id.

File: test/e2e/ipfs-companion.test.js
Comment (galargh) [reply test/e2e/ipfs-companion.test.js:96]: In all honest, the method that lead me to this value wasn't very scientific but it seems to work.

File: test/e2e/ipfs-companion.test.js
Comment (galargh) [reply test/e2e/ipfs-companion.test.js:124]: I like this idea! I'd hold off with it until we have more test cases to cover.

**Comment by SgtPooki** [comment]
I would also like to voice my preference of playwright over selenium. 

1. We're using playwright with all of our projects
2. Selenium is historically unstable during runtime and maintenance. I've got experience with building & maintaining an e2e library backed by selenium, and I am not a fan.

**Comment by galargh** [comment]
> I would also like to voice my preference of playwright over selenium.
> 
> 1. We're using playwright with all of our projects
> 2. Selenium is historically unstable during runtime and maintenance. I've got experience with building & maintaining an e2e library backed by selenium, and I am not a fan.

Personally, I don't know enough about either to prefer one over the other **BUT** it seems we cannot accomplish the task at all in playwright at the moment because of lack of support for Firefox extensions (https://github.com/microsoft/playwright/issues/7297).

File: ci/download-release-artifacts.sh
Comment (galargh) [reply ci/download-release-artifacts.sh:13]: I didn't but that's because I wanted to use a version that's known to work so that I can work on the test without wondering if it's something in the building process that's affecting my results. It might not be the most common workflow though.

**Comment by SgtPooki** [comment]
> it seems we cannot accomplish the task at all in playwright

ACK. makes complete sense.

File: .github/workflows/e2e.yml
Comment (SgtPooki) [inline .github/workflows/e2e.yml:36]: we should be able to use ${env.IPFS_COMPANION_VERSION} in the `if` shouldn't we? I assume `IPFS_COMPANION_VERSION` is required by the script itself (but we could also pass it as an argument)

Comment (SgtPooki) [reply .github/workflows/e2e.yml:36]: ultimately, It would be nice if these two steps comparing `inputs.ipfs-companion-version` could be combined

File: ci/download-release-artifacts.sh
Comment (SgtPooki) [reply ci/download-release-artifacts.sh:13]: I don't think this script would be called locally unless explicitly wanting to test the functionality of downloading the assets from github @whizzzkid. It should only be called by `e2e.yml` when `ipfs-companion-version` is not an empty string, or when you're wanting to validate, locally, that the script pulls from github appropriately.

File: docker-compose.e2e.yml
Comment (SgtPooki) [inline docker-compose.e2e.yml:12]: +1. thanks for this

File: docker-compose.e2e.yml
Comment (SgtPooki) [inline docker-compose.e2e.yml:1]: Note that we should be able to [set hostname and domain name](https://docs.docker.com/compose/compose-file/compose-file-v3/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir) for these containers, which would make addressing https://github.com/ipfs/ipfs-companion/pull/1121/files#r1053853150 much easier. We would know the hostname, protocol, and port

File: test/e2e/ipfs-companion.test.js
Comment (SgtPooki) [reply test/e2e/ipfs-companion.test.js:31]: @galargh linter should not complain about `!= null`. If it is, then the type is overloaded or something else is off, but ditto to what Nishant said. 

I'm fine with `!== undefined` for env vars because they're either strings or undefined, but I prefer `!= null` as a general "maintainable code" practice for communicating to future devs (myself) that it doesn't matter that it's not explicitly undefined, only that it's defined. But again, for env vars, the nuance here is tiny, if not completely non-existent.

File: test/e2e/ipfs-companion.test.js
Comment (SgtPooki) [inline test/e2e/ipfs-companion.test.js:37]: I'm not a fan of hardcoded delays, though I know they're sometimes necessary. We should use something like pRetry instead with some kind of backoff (increasing delays). i.e. we should check if the thing is ready and return success when it is. 

Otherwise, we're saying no matter what improvements happen in the future, all of this code will be at least `sumOfAllDelays + actualRuntime` long. For e2e tests that I imagine we want to run on all PRs in the future.. that's not ideal.

We used to have `window.ipfsCompanion`, but that was removed.. @lidel @whizzzkid are there any other ways to confirm that companion is loaded and ready?

One option: use pRetry with `findExtensionUrl`

File: test/e2e/ipfs-companion.test.js
Comment (SgtPooki) [inline test/e2e/ipfs-companion.test.js:126]: nit: 
```suggestion
async function runBasicBrowserTest (browserName) {
```

non-nit comment: I feel like this test is set up in a way that all of the tests are going to be added here, which we don't want. We can tackle this later.

File: test/e2e/ipfs-companion.test.js
Comment (SgtPooki) [inline test/e2e/ipfs-companion.test.js:130]: IIRC, You're supposed to call `.true` as a function here.

File: test/e2e/ipfs-companion.test.js
Comment (SgtPooki) [inline test/e2e/ipfs-companion.test.js:144]: call the `undefined` function as eslint yells about the syntactic sugar.

File: test/e2e/ipfs-companion.test.js
Comment (SgtPooki) [inline test/e2e/ipfs-companion.test.js:158]: I feel like this will be flaky unless we have a guaranteed peer running in CI as well.

File: test/e2e/ipfs-companion.test.js
Comment (SgtPooki) [inline test/e2e/ipfs-companion.test.js:172]: The test is actually doing much more than this. If we did something like 

```js
await runTest('chromium', async (browser, url) => {
    const peers = await getNumberOfConnectedPeers(browser, url)

    expect(peers).not.to.equal(0)
})
```

then this might be a hair more accurate and help us add tests more easily in the future.

File: test/e2e/ipfs-companion.test.js
Comment (SgtPooki) [reply test/e2e/ipfs-companion.test.js:96]: selenium is full of bad async logic.. delays will continue to be prominent and non-scientific

**Review by SgtPooki — APPROVED** [review]
Approving with comments.. thanks a ton for putting this together @galargh

**Comment by SgtPooki** [comment]
@galargh , the only thing I would really like to see before merge is an update to README.md or docs/DEVELOPER-NOTES.md informing users how to run this locally, how to debug, etc.

**Review by SgtPooki — CHANGES_REQUESTED** [review]
When attempting to run locally, I ran into some issues:

```bash
npm run clean
npm i
npm run release-build # Error when running this
```

### Error output
```bash
 => [3/9] RUN mkdir -p /home/node/app                                                                                                         0.2s
 => ERROR [4/9] RUN if [ 501 -ne 0 ] && [ 20 -ne 0 ]; then     userdel -f node &&     if getent group node ; then groupdel node; fi &&     g  0.3s
------
 > [4/9] RUN if [ 501 -ne 0 ] && [ 20 -ne 0 ]; then     userdel -f node &&     if getent group node ; then groupdel node; fi &&     groupadd -g 20 node &&     useradd -l -u 501 -g node node &&     chown -fhR 501:20 /home/node; fi:
#8 0.255 groupadd: GID '20' already exists
------
executor failed running [/bin/sh -c if [ ${USER_ID:-0} -ne 0 ] && [ ${GROUP_ID:-0} -ne 0 ]; then     userdel -f node &&     if getent group node ; then groupdel node; fi &&     groupadd -g ${GROUP_ID} node &&     useradd -l -u ${USER_ID} -g node node &&     chown -fhR ${USER_ID}:${GROUP_ID} /home/node; fi]: exit code: 4
```

File: .github/workflows/e2e.yml
Comment (SgtPooki) [inline .github/workflows/e2e.yml:39]: we should eventually upload the built docker image (`ipfs-companion-release-build`) and reuse that if possible.. (hash of source or non-docker build artifacts to determine if we want to use the same image?)

**Review by galargh — COMMENTED** [review]
> @galargh , the only thing I would really like to see before merge is an update to README.md or docs/DEVELOPER-NOTES.md informing users how to run this locally, how to debug, etc.

Docs added to docs/DEVELOPER-NOTES.md.

> When attempting to run locally, I ran into some issues.

I added deleting user/group if it already exists to the `Dockerfile` which should fix the issue. I suspect it's there to maintain correct ownership of built files after we get them out of the docker container.

File: .github/workflows/e2e.yml
Comment (SgtPooki) [reply .github/workflows/e2e.yml:12]: so i'm guessing these are more for the workflow_dispatch method. makes sense

File: docker-compose.e2e.yml
Comment (SgtPooki) [reply docker-compose.e2e.yml:1]: Are you asking about how hostname and domainname are set inside the container? I'm not fully aware of the implementation details but I believe the etc/hostname file (and docker configs) are updated appropriately. 

Ultimately, setting domainname and hostname would allow you to address the CORS concerns brought up by @whizzzkid, because you would be able to specify the origins set via hostname/domainname and port.

**Review by SgtPooki — APPROVED** [review]
lgtm..  a few comments unaddressed, but a significant improvement. Thanks @galargh !

File: docker-compose.e2e.yml
Comment (galargh) [reply docker-compose.e2e.yml:1]: I got confused a bit and though these were global settings for the compose file 🤦 

Back to the matter, we're already "using" hostname. If the value is not specified, then the hostname defaults to the service name. We already use this fact to communicate between services. E.g. that's how browsers know they can find Kubo at http://kubo. 

Domainname, I'm not so familiar with, but apparently it's not that useful - https://stackoverflow.com/questions/70741780/what-is-the-domainname-option-used-for-in-docker

Anyway, I tried to be more explicit about access control and limit the origins to `http://chromium` and `http://firefox` but, apparently, that's not enough.

You can now set access control values through env - https://github.com/ipfs/ipfs-companion/blob/25a4760138ac21fdfdf198a6476a88b92bd09c4c/ci/access-control-allow-all.sh. This should allow for easier experimentation in the future.

Comment (galargh) [reply docker-compose.e2e.yml:1]: I created an issue for this at https://github.com/ipfs/ipfs-companion/issues/1130

File: .github/workflows/e2e.yml
Comment (galargh) [reply .github/workflows/e2e.yml:39]: I created https://github.com/ipfs/ipfs-companion/issues/1132 for this

**Pull request merged**
